### PR TITLE
Feat: move injection equipment into windows dll

### PR DESCRIPTION
**Pull request by notnotmelon**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -99,24 +99,90 @@ impl PDBCache {
...
+    static DYNAMIC_LIBRARY_SUFFIX: &str = ".so";
+    #[cfg(target_os = "linux")]
+    static RIVETS_LIB: &str = "rivets.so";
+    #[cfg(target_os = "windows")]
+    static DYNAMIC_LIBRARY_SUFFIX: &str = ".dll";
+    #[cfg(target_os = "windows")]
+    static RIVETS_LIB: &str = "rivets.dll";
+
+    let mut result = vec![];
+    let mod_list = ModList::generate_custom(read_data, &write_data)?;
```
Comment (fgardt) [inline src/lib.rs]: Fixing the issue that caused injector to not find mods. Also run `cargo update` to get an updated version of `mod_util`
```suggestion
    let mut mod_list = ModList::generate_custom(read_data, &write_data)?;
    mod_list.load();
```

File: src/lib.rs [hunk 2]
```diff
@@ -99,24 +99,90 @@ impl PDBCache {
...
+
+        let extracted_lib_name = format!("{factorio_mod_name}{DYNAMIC_LIBRARY_SUFFIX}");
+        let lib_path = write_data
+            .as_ref()
+            .join("temp/rivets")
+            .join(extracted_lib_name);
+        std::fs::write(&lib_path, lib)?;
+
+        result.push(lib_path);
+    }
```
Comment (fgardt) [inline src/lib.rs]: I feel like the `factorio_` prefix is a bit verbose and not needed?

```suggestion
    for mod_name in mod_load_order {
        let current_mod = all_active_mods
            .get(&mod_name)
            .expect("The list of active mods contains all mods in the load order");

        let lib = match current_mod.get_file(RIVETS_LIB) {
            Err(ModError::PathDoesNotExist(_)) => continue,
            Ok(lib) => lib,
            Err(e) => return Err(e.into()),
        };

        std::fs::create_dir_all(write_data.as_ref().join("temp/rivets"))?;

        let extracted_lib_name = format!("{mod_name}{DYNAMIC_LIBRARY_SUFFIX}");
        let lib_path = write_data
            .as_ref()
            .join("temp/rivets")
            .join(extracted_lib_name);
        std::fs::write(&lib_path, lib)?;

        result.push(lib_path);
    }
```

File: src/lib.rs [hunk 3]
```diff
@@ -99,24 +99,90 @@ impl PDBCache {
 /// # Safety
 /// This function is unsafe because it uses the Windows API.
 /// Do not call this function in a threaded context.
-unsafe fn inject(
-    function_name: &str,
-    hook: unsafe fn(u64) -> Result<(), rivets::retour::Error>,
-) -> Result<()> {
+unsafe fn inject(hook: &RivetsHook) -> Result<()> {
     let factorio_path = std::path::Path::new("C:/Users/zacha/Documents/factorio/bin");
```
Comment (fgardt) [inline src/lib.rs]: hardcoded path? :)

File: src/lib.rs [hunk 4]
```diff
@@ -99,24 +99,90 @@ impl PDBCache {
...
-    unsafe { back(this) }
+unsafe fn main(read_path: PathBuf, write_path: PathBuf) -> Result<()> {
+    type ABIWrapper = extern "C" fn() -> *const rivets::RivetsFinalizeABI;
+
+    for dll_so_file in extract_all_mods_libs(read_path, write_path)? {
+        let dll_so_file = Library::new(dll_so_file)?;
+
+        let rivets_finalize_abi: libloading::Symbol<ABIWrapper> =
+            dll_so_file.get(b"rivets_finalize\0")?;
+        let rivets_finalize_abi = &*rivets_finalize_abi();
```
Comment (fgardt) [inline src/lib.rs]: ?

**Review by fgardt — CHANGES_REQUESTED** [review]
I think it might make sense to expand this PR to be fully cross platform? Should not be too much that needs to be moved / annotated with cfg to get it to compile on https://github.com/factorio-rivets/rivets-factorio-mod/labels/os%3A%3Alinux. \
`libloading` and all the parts providing it with binaries & paths are already cross platform so there should not be much work (if any) to get it functional once it actually compiles :)

**Review by fgardt — APPROVED** [review]
LGTM

**Comment by notnotmelon** [comment]
This still needs more work, sorry

**Pull request merged**